File: Sources/HomieProtocol/Paths.swift
import Foundation

/// Filesystem locations and environment variables shared by daemon, app, and CLI.
public enum HomiePaths {
    public static var appSupport: URL {
        FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("Homie", isDirectory: true)
    }

    public static var socket: URL { appSupport.appendingPathComponent("daemon.sock") }
    public static var stateFile: URL { appSupport.appendingPathComponent("state.json") }
    public static var logsDir: URL { appSupport.appendingPathComponent("logs", isDirectory: true) }
    public static var holdersDir: URL {
        appSupport.appendingPathComponent("holders", isDirectory: true)
    }
    public static var injectDir: URL { appSupport.appendingPathComponent("inject", isDirectory: true) }
    public static var binDir: URL { appSupport.appendingPathComponent("bin", isDirectory: true) }
    public static var manifestOverridesDir: URL {
        appSupport.appendingPathComponent("manifests/overrides", isDirectory: true)
    }
    public static var daemonLogFile: URL { appSupport.appendingPathComponent("homied.log") }
    /// Remote-access config (port + token). Present ⇔ remote TCP listener enabled.
    public static var remoteConfigFile: URL { appSupport.appendingPathComponent("remote.json") }
    /// Catalog of SSH-reachable execution hosts. Missing ⇔ Local only.
    public static var hostsConfigFile: URL { appSupport.appendingPathComponent("hosts.json") }

    public static func ensureDirectoriesExist() throws {
        for dir in [appSupport, logsDir, holdersDir, injectDir, binDir, manifestOverridesDir] {
            try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
            try FileManager.default.setAttributes(
                [.posixPermissions: NSNumber(value: Int16(0o700))],
                ofItemAtPath: dir.path)
        }
        // Tighten existing installations as part of every daemon boot. These
        // files can contain terminal output, local paths, and pairing secrets.
        for file in [stateFile, daemonLogFile, remoteConfigFile,
                     appSupport.appendingPathComponent("homied.boot.log"),
                     appSupport.appendingPathComponent("daemon.lock")]
        where FileManager.default.fileExists(atPath: file.path) {
            try? FileManager.default.setAttributes(
                [.posixPermissions: NSNumber(value: Int16(0o600))],
                ofItemAtPath: file.path)
        }
        if let logs = try? FileManager.default.contentsOfDirectory(
            at: logsDir, includingPropertiesForKeys: nil) {
            for file in logs where file.pathExtension == "bin" {
                try? FileManager.default.setAttributes(
                    [.posixPermissions: NSNumber(value: Int16(0o600))],
                    ofItemAtPath: file.path)
            }
        }
    }
}

public enum HomieEnv {
    /// Set in every PTY the daemon spawns; how helpers and MCP identify the caller.
    public static let sessionID = "HOMIE_SESSION_ID"
    /// Path to the daemon control socket.
    public static let socket = "HOMIE_SOCKET"
    /// Path to the `homie` CLI binary (hooks reference it via env, not hardcoded paths).
    public static let cli = "HOMIE_CLI"
}

File: homie/crates/homie-proto/src/paths.rs
//! Pure path construction and shared environment names.
//!
//! This crate deliberately does not discover a home directory or touch the
//! filesystem. Callers provide the user's home directory.

use std::path::{Path, PathBuf};

pub const APP_SUPPORT_RELATIVE_PATH: &str = "Library/Application Support/Homie";
pub const SOCKET_FILE_NAME: &str = "daemon.sock";
pub const STATE_FILE_NAME: &str = "state.json";
pub const LOGS_DIR_NAME: &str = "logs";
pub const INJECT_DIR_NAME: &str = "inject";
pub const BIN_DIR_NAME: &str = "bin";
pub const MANIFEST_OVERRIDES_RELATIVE_PATH: &str = "manifests/overrides";
pub const DAEMON_LOG_FILE_NAME: &str = "homied.log";
pub const HOSTS_CONFIG_FILE_NAME: &str = "hosts.json";

pub const ENV_SESSION_ID: &str = "HOMIE_SESSION_ID";
pub const ENV_SOCKET: &str = "HOMIE_SOCKET";
pub const ENV_CLI: &str = "HOMIE_CLI";

pub struct HomiePaths;

impl HomiePaths {
    pub fn app_support(home: impl AsRef<Path>) -> PathBuf {
        home.as_ref().join(APP_SUPPORT_RELATIVE_PATH)
    }

    pub fn socket(home: impl AsRef<Path>) -> PathBuf {
        Self::app_support(home).join(SOCKET_FILE_NAME)
    }

    pub fn state_file(home: impl AsRef<Path>) -> PathBuf {
        Self::app_support(home).join(STATE_FILE_NAME)
    }

    pub fn logs_dir(home: impl AsRef<Path>) -> PathBuf {
        Self::app_support(home).join(LOGS_DIR_NAME)
    }

    pub fn inject_dir(home: impl AsRef<Path>) -> PathBuf {
        Self::app_support(home).join(INJECT_DIR_NAME)
    }

    pub fn bin_dir(home: impl AsRef<Path>) -> PathBuf {
        Self::app_support(home).join(BIN_DIR_NAME)
    }

    pub fn manifest_overrides_dir(home: impl AsRef<Path>) -> PathBuf {
        Self::app_support(home).join(MANIFEST_OVERRIDES_RELATIVE_PATH)
    }

    pub fn daemon_log_file(home: impl AsRef<Path>) -> PathBuf {
        Self::app_support(home).join(DAEMON_LOG_FILE_NAME)
    }

    pub fn hosts_config_file(home: impl AsRef<Path>) -> PathBuf {
        Self::app_support(home).join(HOSTS_CONFIG_FILE_NAME)
    }
}

pub struct HomieEnv;

impl HomieEnv {
    pub const SESSION_ID: &'static str = ENV_SESSION_ID;
    pub const SOCKET: &'static str = ENV_SOCKET;
    pub const CLI: &'static str = ENV_CLI;
}
